vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -238,15 +238,30 @@ import { isFocusable } from '@vaadin/grid/src/vaadin-grid-active-item-mixin.js';
         });
         grid._createPropertyObserver('activeItem', '__activeItemChangedDetails', true);
 
-        grid.$connector._getPageIfSameLevel = tryCatchWrapper(function (parentKey, index, defaultPage) {
-          let cacheAndIndex = grid._cache.getCacheAndIndex(index);
-          let parentItem = cacheAndIndex.cache.parentItem;
-          let parentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
-          if (parentKey !== parentKeyOfIndex) {
-            return defaultPage;
-          } else {
+        grid.$connector._getSameLevelPage = tryCatchWrapper(function (parentKey, index, defaultPage) {
+          const cacheAndIndex = grid._cache.getCacheAndIndex(index);
+          const parentItem = cacheAndIndex.cache.parentItem;
+          const parentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
+          if (parentKeyOfIndex === parentKey) {
             return grid._getPageForIndex(cacheAndIndex.scaledIndex);
           }
+          return this._getSameLevelPageByHierarchy(parentKey, defaultPage, parentKeyOfIndex);
+        });
+
+        grid.$connector._getSameLevelPageByHierarchy = tryCatchWrapper(function (parentKey, defaultPage, parentKeyOfIndex) {
+          if (!parentKeyOfIndex && parentKeyOfIndex === root) {
+            return defaultPage;
+          }
+          const parentCacheAndIndex = grid._cache.getCacheAndIndexByKey(parentKeyOfIndex);
+          if (!parentCacheAndIndex) {
+            return defaultPage;
+          }
+          const parentItem = parentCacheAndIndex.cache.parentItem;
+          const updatedParentKeyOfIndex = parentItem ? grid.getItemId(parentItem) : root;
+          if (updatedParentKeyOfIndex === parentKey) {
+            return grid._getPageForIndex(parentCacheAndIndex.scaledIndex);
+          }
+          return this._getSameLevelPageByHierarchy(parentKey, defaultPage, updatedParentKeyOfIndex);
```

Comment:
Could add some code comments/documentation to make it easier to read in the future (and review now).

Reply:
Added some brief comments.